### DISCUSSION TITLE
Direct exchange does not route messages when queues are bound with the # pattern

### DISCUSSION THREAD

**Discussion by joeljhou** [rabbitmq/rabbitmq-server#11268] [Other] [answered]
### Describe the bug

When using a custom DirectExchange in RabbitMQ and binding a queue to it with the routing key #, the queue does not receive any messages, even though it should receive messages with any routing key.

### Reproduction steps

1. Create a custom DirectExchange in RabbitMQ.
2. Bind a queue to this exchange using the routing key #.
3. Send a message to the exchange with any routing key.
4. Check if the message is received by the bound queue.

### Expected behavior

**Expected behavior**
The queue should receive messages when binding it to the custom direct exchange with the routing key #, as # should match any routing key.

**Actual behavior**
The queue does not receive any messages when binding it with the routing key #.

### Additional context

**Example code**
```
@Test
public void testSendMessageAndReceiveMessage() {
    // Declare the exchange
    rabbitAdmin.declareExchange(new DirectExchange("my.direct.exchange", true, false));

    // Declare the queue
    rabbitAdmin.declareQueue(new Queue("test.direct.queue", true));

    // Bind the queue to the custom direct exchange
    Binding binding = new Binding(
            "test.direct.queue", Binding.DestinationType.QUEUE,
            "my.direct.exchange", "#", null);
    rabbitAdmin.declareBinding(binding);

    // Send a message to the custom direct exchange
    rabbitTemplate.convertAndSend("my.direct.exchange", "random", "Hello, RabbitMQ!");

    // Receive the message
    Object directMessage = rabbitTemplate.receiveAndConvert("test.direct.queue");
    System.out.println("directMessage = " + directMessage);
}
```

This code snippet demonstrates the issue where the queue does not receive messages when binding it to the custom direct exchange with routing key #.

Here's the version information you provided:

* RabbitMQ Java Client: 5.21.0
* Spring Rabbit: 3.1.4

**Comment by joeljhou** [comment]
<img width="1440" alt="image" src="https://github.com/rabbitmq/rabbitmq-website/assets/45135095/0732800c-a272-4f99-b2e5-dcdcbe2a4312">

Results that do not meet expectations
```
directMessage = null
```

**Reply by michaelklishin** [reply]
You have chosen the wrong exchange type. You need a [topic exchange](https://www.rabbitmq.com/tutorials/tutorial-five-spring-amqp) and not a [direct one](https://www.rabbitmq.com/tutorials/tutorial-four-spring-amqp).

**Comment by michaelklishin** [comment, marked as answer]
@joeljhou your expectations are not correct. Direct exchanges do not use patterns, topic exchanges do. See [tutorial 5](https://www.rabbitmq.com/tutorials/tutorial-five-spring-amqp).

Direct exchanges technically can route to multiple queues if they are bound with the same exact routing key but they do not and never have supported patterns.
